### DISCUSSION TITLE
Is there a good way to pair outputs with inputs ?

### DISCUSSION THREAD

**Discussion by vexcited** [djipco/webmidi#280] [Q&A]
Hello !

I'm currently developing an application where a user can select a MIDI device and do things with it
and I have multiple questions about it...

- Can two connected MIDI devices have the exact same name ?
- How can we pair the outputs with the outputs of these devices ?

Currently I was neutralizing the IN and OUT of the names and pairing them using the names of the inputs/outputs.

But I was asking myself if at some point, two devices have the exact same name, how it would it work ?

Is there an official way for pairing inputs/outputs or the name match method is the only way to go ?

Thanks you !

**Comment by djipco** [comment]
> * Can two connected MIDI devices have the exact same name ?

I have never tried it but I don't see any reason why that wouldn't possible. The names will be the same but the IDs will be different.

> * How can we pair the outputs with the outputs of these devices ?

I guess you mean "_pair the **inputs** with the outputs_"? If that's the case, the Web MIDI API currently falls short. There is a [proposal](https://github.com/WebAudio/web-midi-api/issues/148) to add a `MIDIDevice` object to address this problem. But, it's scheduled for v2 of the spec which is nowhere near being adopted. 

Do you have two devices of the same kind? I would be curious to know how they are reported at the system level and at the Web MIDI API level. If you do, could you use the code below to print out the info and post it in this thread?

```js
console.log(WebMidi.getOutputByName("name here")._midiOutput);
console.log(WebMidi.getInputByName("name here")._midiInput);
```

> Is there an official way for pairing inputs/outputs or the name match method is the only way to go ?

Currently, the name match method is the way to go.

Until an eventual update of the [Web MIDI API spec](https://webaudio.github.io/web-midi-api/) is released, maybe this is something that could be integrated within WEBMIDI.js itself (using the name match approach). Basically, the `Input` and `Output` objects would have a `device` property pointing to a `Device` object. The `Device` object would have an `id` and a `name` property. It would also have lists of associated inputs (`inputs`) and outputs (`outputs`). We may be able to use an [identity request](http://midi.teragonaudio.com/tech/midispec/identity.htm) sysex message to figure out which ports belong to which device in the case of identical devices.

**Reply by vexcited** [reply]
Thanks you for the reply !

Unfortunately I don't have two devices of the same kind but if I'm asking is because something weird happened to a friend of mine who was on Chrome Android and connected one device and WebMIDI just gave two outputs with the exact same name, [here is a screenshot](https://media.discordapp.net/attachments/989952883132882984/994556474007228447/Screenshot_20220707-125153_Chrome.jpg). I will try to ask him to retry with Kiwi Browser so he could show me more logs about it.

That `MIDIDevice` proposal looks promising though.

**Reply by djipco** [reply]
> a friend of mine who was on Chrome Android and connected one device and WebMIDI just gave two outputs with the exact same name, [here is a screenshot](https://media.discordapp.net/attachments/989952883132882984/994556474007228447/Screenshot_20220707-125153_Chrome.jpg).

It's hard to say to happened. What was the content of `WebMidi.outputs` ?

**Reply by vexcited** [reply]
Sorry for being late on this one. I had to actually buy a OTG cable and test it myself so I can give you what I had.

> What was the content of `WebMidi.outputs` ?

Here you go, and even on my Android device, I had the same name issue.

| `WebMidi.outputs[0]` | `WebMidi.outputs[1]` |
| ---------------------- | ---------------------- |
| ![output-0](https://user-images.githubusercontent.com/59152884/181115852-4f58d098-5797-4d44-94df-64d0d3efbae0.jpg) | ![output-1](https://user-images.githubusercontent.com/59152884/181115855-b419c8c6-2e78-4ed3-b213-4fb920097581.jpg) |

**Reply by vexcited** [reply]
So, it seems like that in Android the ID actually means something I guess ?

Because on PC we have like some numbers, but here, we have the index of the port ?

The ID for the outputs were `native:port-out-0` and `native:port-out-1` and the inputs were `native:port-in-2` and `native:port-in-3`.

So we can actually match them by doing like `output[index]` matchs with `input[index]` I guess ? But I don't know if that's a good idea actually lol

If that works, it would always work ; but I don't know if the inputs are in the same order than the outputs (here it seems yes but in real world, I don't know)

**Comment by djipco** [comment]
I added the idea of including a `Device` object to the [v3 roadmap](https://webmidijs.org/docs/roadmap/v3).

**Comment by rubendax** [comment]
> Do you have two devices of the same kind? I would be curious to know how they are reported at the system level and at the Web MIDI API level. If you do, could you use the code below to print out the info and post it in this thread?
> 
>      console.log(WebMidi.getOutputByName("name here")._midiOutput);
>      console.log(WebMidi.getInputByName("name here")._midiInput);

**Two Keith McMillen BopPads simultaneously USB connected** (Mac OS 10.15.7, Electron 12.0.5):

It seems that this code only prints the information for one of the devices:
```js
console.log(WebMidi.getOutputByName("BopPad")._midiOutput);
```
![image](https://user-images.githubusercontent.com/3331057/185062612-a3c7eae4-595c-49a8-a1c1-2cdb6fc55a10.png)

If instead I listen for all ```connected``` events then the console shows information about both devices:
```js
WebMidi.addListener("connected", midiDevice => {
        if (midiDevice.port.name == "BopPad") {
            if (midiDevice.port.type == "output") {
                console.log(midiDevice.port._midiOutput);
            }
        }
    });
```
![image](https://user-images.githubusercontent.com/3331057/185060519-3f5e1405-a04b-4d56-a8c8-4b7874b4135c.png)

Let me know if you want me to check any other duplicated device behavior.

**Reply by djipco** [reply]
Thanks for reporting! As I expected, the devices have the same name and manufacturer but have unique IDs. I guess using the name match approach should suffice for most use cases...
